benchmark/sparse_blas/operations.cpp

```diff
@@ -0,0 +1,569 @@
+    {
+        auto ref = gko::ReferenceExecutor::create();
+        auto correct = gko::clone(ref, mtx2_.get());
+        gko::clone(ref, mtx_)->apply(scalar_.get(), id_.get(), scalar_.get(),
+                                     correct.get());
+        return validate_result(correct.get(), gko::clone(ref, mtx_out_).get());
+    }
+
+    gko::size_type get_flops() const override
+    {
+        return mtx_->get_num_stored_elements() +
+               mtx2_->get_num_stored_elements();
+    }
+
+    gko::size_type get_memory() const override
+    {
+        // read and write everything only once, ignore row pointers
+        return (mtx_->get_num_stored_elements() +
+                mtx2_->get_num_stored_elements() +
+                mtx_out_->get_num_stored_elements()) *
+               (sizeof(etype) + sizeof(itype));
+    }
+
+    void prepare() override { mtx_out_ = mtx2_->clone(); }
+
+    void run() override
+    {
+        mtx_->apply(scalar_.get(), id_.get(), scalar_.get(), mtx_out_.get());
+    }
+
+private:
+    const Mtx* mtx_;
+    std::unique_ptr<Mtx> mtx2_;
+    std::unique_ptr<gko::matrix::Dense<etype>> scalar_;
+    std::unique_ptr<gko::matrix::Identity<etype>> id_;
+    std::unique_ptr<Mtx> mtx_out_;
+};
+
+
+class TransposeOperation : public BenchmarkOperation {
+public:
+    explicit TransposeOperation(const Mtx* mtx) : mtx_{mtx} {}
+
+    std::pair<bool, double> validate() const override
+    {
+        auto ref = gko::ReferenceExecutor::create();
+        return validate_result(
+            gko::as<Mtx>(gko::clone(ref, mtx_)->transpose()).get(),
+            gko::clone(ref, mtx_out_).get());
+    }
+
+    gko::size_type get_flops() const override { return 0; }
+
+    gko::size_type get_memory() const override
+    {
+        // read and write everything only once, ignore row pointers
+        return 2 * mtx_->get_num_stored_elements() *
+               (sizeof(etype) + sizeof(itype));
+    }
+
+    void prepare() override { mtx_out_ = nullptr; }
+
+    void run() override { mtx_out_ = gko::as<Mtx>(mtx_->transpose()); }
+
+private:
+    const Mtx* mtx_;
+    std::unique_ptr<Mtx> mtx_out_;
+};
+
+
+class SortOperation : public BenchmarkOperation {
+public:
+    explicit SortOperation(const Mtx* mtx)
+    {
+        mtx_shuffled_ = mtx->clone();
+        gko::test::unsort_matrix(mtx_shuffled_.get(), get_engine());
+        mtx_out_ = mtx_shuffled_->clone();
+    }
+
+    std::pair<bool, double> validate() const override
+    {
+        auto ref = gko::ReferenceExecutor::create();
+        auto mtx_sorted = gko::clone(ref, mtx_shuffled_);
+        mtx_sorted->sort_by_column_index();
+        return validate_result(mtx_sorted.get(),
+                               gko::clone(ref, mtx_out_).get());
+    }
+
+    gko::size_type get_flops() const override { return 0; }
+
+    gko::size_type get_memory() const override
+    {
+        // read and write everything only once, read row pointers once
```

Comment:
I feel this is a bit misleading. We use black box algorithms to sort each row, so we can't say with certainty how many reads we are actually doing. I would prefer if this function returns some invalid value.

Reply:
I think for this kind of segmented sorting, assuming a simple model (no extremely long rows, infinitely fast local memory), I would consider this a fairly good approximation - load the data once (into cache or shared memory doesn't matter much), sort it locally and store the sorted results back. For sorting algorithms, common metrics usually involve either elements/s or runtime/n log n (as the theoretical optimum), the elements/s metric is pretty close to what we are measuring here.